truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentableFactory.java

```diff
@@ -27,49 +27,80 @@
 import com.oracle.truffle.api.nodes.Node;
 
 /**
- * Factory capable to create instance of a wrapper node for provided node. Usually there is no need
- * to implement this factory manually - use {@link Instrumentable} annotation and it will generate
- * the factory for you.
- * 
- * @param <T> the type of node this factory operates on
+ * Factory for creating <em>wrapper nodes</em>. The instrumentation framework inserts a wrapper
+ * between an {@link Instrumentable} guest language node (called the <em>delegate</em>) and its
+ * parent for the purpose of interposing on execution events at the delegate and reporting those
+ * events to the instrumentation framework.</p>
+ * <p>
+ * Wrapper implementations can be generated automatically: see the {@link Instrumentable}
+ * annotation.
+ *
+ * @param <T> the type of delegate node this factory operates on
  * @since 0.12
  */
 public interface InstrumentableFactory<T extends Node> {
 
     /**
+     * Returns a new, never adopted, unshared {@link WrapperNode wrapper} node implementation for a
+     * particular {@link Instrumentable} node of the guest language AST called its <em>delegate</em>
+     * . The returned wrapper implementation must extend the same type that is annotated with
+     * {@link Instrumentable}. </p>
+     * <p>
+     * A wrapper forwards the following events concerning the delegate to the given
+     * {@link ProbeNode probe} for propagation through the instrumentation framework, e.g. to
+     * {@linkplain ExecutionEventListener event listeners} bound to this guest language program
+     * location:
+     * <ul>
+     * <li>{@linkplain ProbeNode#onEnter(com.oracle.truffle.api.frame.VirtualFrame) onEnter(Frame)}:
+     * an <em>execute</em> method on the delegate is ready to be called;</li>
+     * <li>{@linkplain ProbeNode#onReturnValue(com.oracle.truffle.api.frame.VirtualFrame, Object)
+     * onReturnValue(Frame,Object)}: an <em>execute</em> method on the delegate has just returned a
+     * (possibly <code>null</code>) value;</li>
+     * <li>
+     * {@linkplain ProbeNode#onReturnExceptional(com.oracle.truffle.api.frame.VirtualFrame, Throwable)
+     * onReturnExceptional(Frame,Throwable)}: an <em>execute</em> method on the delegate has just
+     * thrown an exception.</li>
+     * </ul>
+     * </p>
      * <p>
-     * Returns a new, never adopted, unshared instrumentable {@link WrapperNode wrapper} node
-     * implementation for a particular node of the guest language AST. The returned wrapper
-     * implementation must extend the same type that is annotated with {@link Instrumentable} and
-     * forward all events to the given {@link ProbeNode probe}.
+     * A wrapper must pass through a {@link KillException} or {@link QuitException} without any
```

Comment:
CH5) If that is true we need to change the wrapper generation as well.
Currently the generated wrapper pattern looks like this

```
 @Override
        public Object executeGeneric(VirtualFrame frame) {
            try {
                probeNode.onEnter(frame);
                Object returnValue = delegateNode.executeGeneric(frame);
                probeNode.onReturnValue(frame, returnValue);
                return returnValue;
            } catch (Throwable t) {
                probeNode.onReturnExceptional(frame, t);
                throw t;
            }
        }
```

Why is it important that onReturnExceptional is not invoked in case of KillException? I'd say that onReturnValue or onReturnExceptional is **always** invoked even if a KillException was thrown. This would make this specification simpler, because the wrapper can treat any exception the same way.